cpp/dolfinx/fem/utils.h

```diff
@@ -448,15 +448,30 @@ pack_coefficients(const fem::Form<T>& form)
 template <typename T>
 Eigen::Array<T, Eigen::Dynamic, 1> pack_constants(const fem::Form<T>& form)
 {
-  std::vector<T> constant_values;
-  for (auto& constant : form.constants())
+  const auto& constants = form.constants();
+
+  // Calculate size of array needed to store packed constants.
+  Eigen::Index size
+      = std::accumulate(constants.begin(), constants.end(), 0,
+                        [&](Eigen::Index sum, const auto& constant) {
+                          return sum + constant->value.size();
+                        });
+
+  // Pack constants.
+  Eigen::Array<T, Eigen::Dynamic, 1> constant_values(size);
+  Eigen::Index offset = 0;
+  for (const auto& constant : constants)
   {
-    const std::vector<T>& array = constant->value;
-    constant_values.insert(constant_values.end(), array.begin(), array.end());
+    const auto& value = constant->value;
+    const Eigen::Index value_size = value.size();
```

Comment:
I think the below loop could be just
```c++
constant_values.segment(offset, value.size()) = value;
```

However constant data will never be performance critical, so whatever is more readable is preferred.

Reply:
`value` is concretely a `std::vector` so it's not possible. We can discuss if the underlying storage of `function::Constant` should be an `Eigen::Array`, it seems a bit out of place with the current design.

Reply:
Ah ok, I see. Anyway, not too important.